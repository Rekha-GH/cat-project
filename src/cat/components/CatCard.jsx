import React from "react";
import { useSelector } from "react-redux";
import { Card } from "../../common/components";
import {
  setToFavourite,
  deleteFromFavourite,
  changeImageVote,
} from "../model/ImageModel";
import { getCatById } from "../Store/CatSelector";

const CatCard = React.memo((props) => {
  const catObj = useSelector((state) => getCatById(state, props.id));
  const { id, url, likeId, totalVote, voteIds } = catObj;
  const like = likeId !== undefined;

  const likedClicked = React.useCallback(() => {
    if (!like) {
      setToFavourite(id);
    } else {
      deleteFromFavourite(likeId, id);
    }
  }, [like, id, likeId]);

  const clickedUpVote = React.useCallback(() => {
    let value = totalVote + 1;
    changeImageVote(id, value, voteIds);
  }, [id, totalVote, voteIds]);

  const clickedDownVote = React.useCallback(() => {
    let value = totalVote === 0 ? 0 : totalVote - 1;
    changeImageVote(id, value, voteIds);
  }, [id, totalVote, voteIds]);

  return (
    <Card
      liked={likeId !== undefined}
      likeClicked={likedClicked}
      imageUrl={url}
      clickedUpVote={clickedUpVote}
      clickedDownVote={clickedDownVote}
      totalVote={totalVote}
    />
  );
});

export default CatCard;
