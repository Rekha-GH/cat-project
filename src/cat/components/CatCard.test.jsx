import Enzyme, { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";

import * as ReactRedux from "react-redux";
import CatCard from "./CatCard";

Enzyme.configure({ adapter: new Adapter() });

describe("CatCard Test", () => {
  const useSelectorMock = jest.spyOn(ReactRedux, "useSelector");

  jest.spyOn(ReactRedux, "useDispatch");

  test("CatCard Call Correctly", () => {
    useSelectorMock.mockReturnValue({
      id: 1,
      url: "img.jpg",
      likeId: 2845857,
      totalVote: 2,
      voteIds: [],
    });
    const wrapper = shallow(<CatCard />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
