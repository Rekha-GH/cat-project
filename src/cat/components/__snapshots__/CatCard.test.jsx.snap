// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatCard Test CatCard Call Correctly 1`] = `
<Memo()
  clickedDownVote={[Function]}
  clickedUpVote={[Function]}
  imageUrl="img.jpg"
  likeClicked={[Function]}
  liked={true}
  totalVote={2}
/>
`;
